feat: countdown as a Quarto shortcode extension!

coatless opened the pull request:
TODO:

- [x] Rebase (really a big merge commit)
- [x] Add `_extension`
- [x] Re-write R logic into Lua
   - [x] Setup default options where possible
   - [x] ~Mirror whisker write for CSS~
   - [x] Figure whether an option is present and if not use a default
       - Developed `getOption(options, key, default)` for check/retrieval/or use default and `tryOption()` for just check/retrieval
   - [x] Re-work how CSS values are set up
   - [x] ~~Think about options for `prismatic()`~~
- [x] Determine document-level settings

```yaml
countdown:
   font_size: "3rem"
```

- [ ] Figuring out asset sharing in the monorepo: either duplicate libraries via a script

```sh
.
|- lib/
|- quarto/
|  |- _extension/countdown
|- r/
|- python/
|_ README.md
```

c.f. https://github.com/quarto-dev/quarto-cli/discussions/8353

Non-ported functions: 

- [`prismatic::clr_lighten()`](https://github.com/EmilHvitfeldt/prismatic/blob/d1e0b1b5dfbd64cfe3b9e81b1b43ed9af0a069b7/R/lightness.R#L45) (Show stopper)
- [`prismatic::clr_darken()`](https://github.com/EmilHvitfeldt/prismatic/blob/d1e0b1b5dfbd64cfe3b9e81b1b43ed9af0a069b7/R/lightness.R#L142-L144) (depends on `clr_lighten()`)
- [`prismatic::best_contrast()`](https://github.com/EmilHvitfeldt/prismatic/blob/d1e0b1b5dfbd64cfe3b9e81b1b43ed9af0a069b7/R/contrast_ratio.R#L65-L68)

gadenbuie commented:
I don't mean to backseat drive on this, but are you planning on adding (or leaving room for) a python port as well? I could get behind that idea!

But if not, then we might not need to move everything around quite as much and we could let the R package deps in `inst/countdown` be the source of truth. I'm okay with needing to copy from `inst/countdown` into `_extensions/countdown` -- since `quarto add` pulls from main by default it's reasonable to want to be intentional about what's being released into the quarto extension.

coatless commented:
@gadenbuie yup; I'm aiming for a `py` port. If anything, I think it'll be a nice `py-shiny`/`py-shinylive` example

gadenbuie commented:
@coatless I added you as a collaborator in this repo, welcome! Would you like to review #36 before I merge? It prepares the CSS for use in other projects and removes the weird whisker templating. It also sets up the R interface to be closer to what will likely become the Python API. And finally it means you'd have to re-do some of the work you've already done in this PR (sorry! but I think it's worth it to do this now).

gadenbuie commented:
@coatless In terms of logistics, I think we should merge this restructuring as early as possible, I guess ideally after merging the two PRs I put up today.

I'm okay with main being a little messy for a while (we're protected by the CRAN release). Merging the big re-org will let us collaborate on various areas to move this all forward and will let us submit smaller PRs. Minimally I think we should:

1. Re-organize the folder structure
2. Update the docs for installing the R package from GitHub
3. Have CI working for the R package folder
4. Everything else can be stubbed or in a semi broken state to be fixed or implemented in subsequent PRs

coatless commented:
@gadenbuie thanks for the invite! I'm not too worried about the re-work. By moving away from `mustache`-like templating, that'll save one dependency for R (`whiskers`) and Python (`chevron`). 

How would you feel if I split off the restructuring into a separate PR and, then, rebased this PR ontop of the restructured repository?

gadenbuie commented:
> How would you feel if I split off the restructuring into a separate PR and, then, rebased this PR ontop of the restructured repository?

That sounds great!

coatless commented:
@gadenbuie we do have a design choice to think of when it comes to the document-level CSS, e.g. `countdown_style()`.

How should the styles be set in document-level settings?

Option A: Top level option `countdown` with styles nested under `styler`

```yaml
countdown:
  styler: 
     font_size: "3rem"
```

Option B: Only allow for top-level options; may or may not mix with other permanent settings.

```yaml
countdown:
   font_size: "3rem"
```

gadenbuie commented:
Should we allow globally setting options like `warn_when` and `update_every` via the YAML frontmatter? Or just style?

I'm of two minds in terms of the YAML structure for the options. On the one hand, with YAML we can be a bit more expressive than I was with the R function options and we could use the list structure to organize options:

```yaml
countdown:
  position:
    bottom: 0
    right: 0,
  play_sound: false
  warn_when: 0
  update_every: 1
  blink_colon: true
  start_immediately: false
  style:
	font_size: 
    margin:
	padding:
	box_shadow:
    border_width:
    border_radius:
    line_height:
    color:
      initial:
	    border:
        background:
        text:
      running:
	    border:
        background:
        text:
      finished:
	    border:
        background:
        text:
      warning:
	    border:
        background:
        text:
```

Alternatively, given that everything is already flattened in the R function signature, I think I might prefer to just have all options under `countdown` and to have their names match with the R parameter names:

```yaml
countdown:
  play_sound: false
  bottom: 0
  right: 0
  top: null
  left: null
  warn_when: 0L
  update_every: 1L
  blink_colon: false
  start_immediately: false
  font_size: null
  margin: null
  padding: null
  box_shadow: null
  border_width: null
  border_radius: null
  line_height: null
  color_border: null
  color_background: null
  color_text: null
  color_running_background: null
  color_running_border: null
  color_running_text: null
  color_finished_background: null
  color_finished_border: null
  color_finished_text: null
  color_warning_background: null
  color_warning_border: null
  color_warning_text: null
```

coatless commented:
> Should we allow globally setting options like warn_when and update_every via the YAML frontmatter? Or just style?

My thought here was to make available settings when possible globally to avoid redefining in long-running presentations. But, maybe we just add that as a feature if an issue ticket comes along? 

>  I think I might prefer to just have all options under countdown and to have their names match with the R parameter names
Okay, we'll aim for Option B then to keep consistency across the packages.

coatless commented:
Okidokie, I've made global options work if the meta key `countdown` is detected.

<img width="2240" alt="Testing" src="https://github.com/gadenbuie/countdown/assets/833642/b5c28ff0-b869-4c23-9d50-184c1a28c4a2">

You can see some options being discussed in the README. Let me know if you have thoughts on structuring.

Still need to:

- rework inline style options to allow for more customization. 
- figure out why the inline style of `bottom: 0; right: 0` is not placing the countdown timer in the bottom right corner within a `revealjs` slide deck.
- figure out porting of `prismatic`.
   - Does this need to be in the first version of the quarto extension?

gadenbuie commented on quarto/_extensions/countdown/assets/countdown.css:
```diff
@@ -0,0 +1,169 @@
+.countdown {
+  --_running-color: var(--countdown-color-running-text, rgba(0, 0, 0, 0.8));
```
Revealjs has different slide scaling mechanism, so I want to change the default margin in revealjs land down to zero. The next three suggestions should handle that.

```suggestion
.countdown {
  --_margin: 0.6em;
  --_running-color: var(--countdown-color-running-text, rgba(0, 0, 0, 0.8));
```

coatless replied:
Should we try to split this out into a subclass? `.reveal .countdown` ?

gadenbuie commented on quarto/_extensions/countdown/assets/countdown.css:
```diff
@@ -0,0 +1,169 @@
+  border-style: solid;
+  border-radius: var(--countdown-border-radius, 0.9rem);
+  box-shadow: var(--countdown-box-shadow, 0px 4px 10px 0px rgba(50, 50, 50, 0.4));
+  margin: var(--countdown-margin, 0.6em);
```
```suggestion
  margin: var(--countdown-margin, var(--_margin, 0.6em));
```

gadenbuie commented on quarto/_extensions/countdown/assets/countdown.css:
```diff
@@ -0,0 +1,169 @@
+.countdown.running:focus-within .countdown-controls>button:active {
+  --_button-bump: 0;
+}
+
```
```suggestion

/* ---- Quarto Reveal.js ---- */
.reveal .countdown {
  --_margin: 0;
}

```

gadenbuie commented on .gitignore:
```diff
@@ -5,7 +5,7 @@
 
 # Directories that start with _
-_*/
+#_*/   # Unblock to allow for Quarto.  
 !_snaps/
```
```suggestion
_dev/
```

This is really the folder I was targeting anyway, so let's just ignore that one. (I have a convention of always using `_dev/` as a scratchpad, place for notes, reprex files, etc.)

gadenbuie replied:
I have local TimeMachine backups and keep some of my work in Dropbox... but yeah, anything important in `_dev/` I try to commit or push somewhere else.

gadenbuie commented on lib/countdown.js:
```diff
@@ -274,6 +274,7 @@ class CountdownTimer {
     }
   }
 
+
```
Was this intentional?

coatless replied:
Addressed in 

https://github.com/gadenbuie/countdown/pull/35/commits/e0ac982d57a17fb3f3229d50e4ba24b838751259

gadenbuie commented on quarto/_extensions/countdown/assets/countdown.css:
```diff
@@ -0,0 +1,169 @@
+.countdown {
+  --_running-color: var(--countdown-color-running-text, rgba(0, 0, 0, 0.8));
+  --_running-background: var(--countdown-color-running-background, #43AC6A);
+  --_running-border-color: var(--countdown-color-running-border, rgba(0, 0, 0, 0.1));
+  --_finished-color: var(--countdown-color-finished-text, rgba(0, 0, 0, 0.7));
```
I started thinking about how we'd replace prismatic and these lines kind of hint at where my head is at.

First, we definitely do not need a prismatic replacement in the lua filter for this PR; we can certainly iterate after merging this one.

The set up with these lines is that we could tint or shade the countdown border and text color based on the background color. I went with shading (darkening) in the CSS, but we could have tinted (lightend) as well with `rgba(255, 255, 255, 0.1)` or similar.

The question is how we'd know whether we should lighten or darken. But I'm okay waiting on that as well and working it out in follow up PRs.

coatless replied:
Thinking out loud, there would be a Python port coming next. So, would either of these options be acceptable to avoid moving it into the Lua filter?

1. Write the lighting and darkening logic in JavaScript (preferred) 
2. Custom CSS properties using `calc` alongside a flag on the inline element? `light-color`/`dark-color`? 

```
:root {
      --main-background-color: #3498db;
      --text-color: calc(1 - var(--main-background-color) * 0.7); /* Default text color */
    }
```

gadenbuie replied:
Yeah, I'd really like to handle it in CSS, because that will be the easiest in the long term. Hopefully sometime soon we'll get [color-contrast()](https://developer.mozilla.org/en-US/docs/Web/CSS/color_value/color-contrast).

For now, I'm okay with the current approach and we can work it out in future iterations.

gadenbuie reviewed (commented):
This looks great! I think I've answered your questions in the comments below, and I've added a few wrinkles:

1. I'd like to support `{{< countdown 3:30 >}}` as an option for specifying the time. I've suggested changes that make this possible without too much hassle. (Also more behind reasoning the comments.)
2. I proposed a small logic change to use the default `bottom=0` value only if neither `top` nor `bottom` were specified. (Same for `right` and `left`.) This one doesn't come with code suggestions, sorry!

coatless commented on quarto/example.qmd:
```diff
@@ -0,0 +1,40 @@
+:::
+
+
+# Default timer
```
Ahh! Thanks for catching this; I thought something more sinister was wrong.

coatless commented on quarto/_extensions/countdown/countdown.lua:
```diff
@@ -0,0 +1,317 @@
+  
+  -- Retrieve named time arguments and fallback on default values if missing
+  local minutes = tonumber(getOption(kwargs, "minutes", 1))
+  local seconds = tonumber(getOption(kwargs, "seconds", 0))
```
Nice! I'll take care of the TODO.

coatless replied:
Revised to issue a warning in: 

https://github.com/gadenbuie/countdown/pull/35/commits/2dd2da96af33abcef0c7061229dffb1b7cf13068

Added guidance to the error message regarding the string fault in

https://github.com/gadenbuie/countdown/pull/35/commits/ac8949704949bc899550e87ee39b0233c0cb882a

gadenbuie replied:
Nice, thanks for adding those in!

coatless commented on quarto/README.md:
```diff
@@ -20,13 +20,94 @@ This command will download and install the extension under the `_extensions` sub
+
+| Style Option  | Default Value                | Description                                                               |
+| ------------- | ---------------------------- | ------------------------------------------------------------------------- |
+| `top`         | `""` (empty)                 | Top position of the HTML element.                                         |
+| `right`       | `"0"`                        | Right position of the HTML element.                                       |
+| `bottom`      | `"0"`                        | Bottom position of the HTML element.                                      |
+| `left`        | `""` (empty)                 | Left position of the HTML element.                                        |
```
Ewk! Not ideal. Let me address that.

coatless replied:
I took care of this in: https://github.com/gadenbuie/countdown/pull/35/commits/e737fea401780117e3818594dd03ad7ffbd0f4af

Though, I'm defaulting back to the bottom right to match with R's version.

gadenbuie replied:
Thanks! I left a comment below, it needs one more little tweak to be in line with the R logic

gadenbuie commented on quarto/_extensions/countdown/countdown.lua:
```diff
@@ -0,0 +1,353 @@
+  local play_sound = tryPlaySound(getOption(kwargs, "play_sound", "false"))
+
+  -- Check to see if positional outcomes are set; if not, default both bottom and right to 0.
+  if isVariableEmpty(tryOption(kwargs, "top"))   and
+     isVariableEmpty(tryOption(kwargs, "left"))  and
+     isVariableEmpty(tryOption(kwargs, "right")) and
+     isVariableEmpty(tryOption(kwargs, "bottom")) then
+    kwargs["bottom"] = 0 
+    kwargs["right"] = 0 
+  end
```
```suggestion
  if isVariableEmpty(tryOption(kwargs, "top")) and
     isVariableEmpty(tryOption(kwargs, "bottom")) then
    kwargs["bottom"] = 0
  end    
     
  if isVariableEmpty(tryOption(kwargs, "left"))  and
     isVariableEmpty(tryOption(kwargs, "right")) then
    kwargs["right"] = 0 
  end
```

This better follows the R logic which is hidden in the function signature: https://github.com/gadenbuie/countdown/blob/34204099a9b575b72504523f739a51c200bbca5f/r/R/countdown.R#L151-L152

This makes it a lot easier to move the timer around:

```
<!-- bottom right -->
{{< countdown 1:30 >}}

<!-- bottom left -->
{{< countdown 1:30 left=0 >}}

<!-- top right -->
{{< countdown 1:30 top=0 >}}

<!-- top left -->
{{< countdown 1:30 top=0 left=0 >}}
```

gadenbuie commented on quarto/_extensions/countdown/countdown.lua:
```diff
@@ -0,0 +1,353 @@
+  color_background = "inherit",
+  -- Text color of the countdown element
+  color_text = "inherit",
+  -- Background color when the countdown is running
+  color_running_background = "#43AC6A",
+  -- Border color when the countdown is running
+  color_running_border = "#2A9B59FF", -- Needs color_darken()
+  -- Text color when the countdown is running
+  color_running_text = 'inherit',
+  -- Background color when the countdown is finished
+  color_finished_background = "#F04124",
+  -- Border color when the countdown is finished
+  color_finished_border = "#DE3000FF",  -- Needs color_darken()
+  -- Text color when the countdown is finished
+  color_finished_text = 'inherit',
+  -- Background color when the countdown has a warning
+  color_warning_background = "#E6C229",
+  -- Border color when the countdown has a warning
+  color_warning_border = "#CEAC04FF", -- Needs color_darken()
+  -- Text color when the countdown has a warning
+  color_warning_text = 'inherit',
+  -- Selector for the countdown element
```
One last color-related thing: for now let's have these defaults be set by the CSS. I set up the CSS so that users can give a running/finished/warning background color and get reasonable text and border colors -- as long as those background colors as light enough.

In fact, now that I'm thinking about it and looking at this section, I think we might want to simplify the implementation a bit to not have `default_style`. The idea would be that all defaults live in the CSS so that the Quarto extension only needs to overlay user updates to those styles.

coatless replied:
@gadenbuie okay; this can be simplified up to just listing variable names. The filter will need a table of parameter names to parse through (c.f. `safeStyle()`). We may want to simplify `countdown_style()` as well since it will automatically populate with all the function defaults.

<img width="937" alt="Screenshot showing the outcome of running `countdown::countdown_style()`" src="https://github.com/gadenbuie/countdown/assets/833642/d251c74c-98c5-4868-a1c3-008c79e9aea5">

gadenbuie replied:
Good point, thanks for bringing this up. I'll have to think about this, on the one hand I'm okay with `countdown_style()` in R having different behavior from the Quarto extension. On the other hand, I wasn't fully intending for every property to be set in every use of `countdown_style()`.

gadenbuie commented on quarto/_extensions/countdown/countdown.lua:
```diff
@@ -0,0 +1,353 @@
+  local minutes, seconds
+
+  -- Retrieve named time arguments and fallback on default values if missing
+  arg_time = parseTimeString(args)
```
I think I missed a local here

```suggestion
  local arg_time = parseTimeString(args)
```

coatless replied:
One of the downsides to Lua; everything is global without using `local` 😞

gadenbuie reviewed (commented):
Things are looking great in here! Thanks for addressing those comments, I've added just a couple more and then this will be ready to go!

coatless commented:
@gadenbuie I re-worked the default arguments so that we're no longer using a table of named entries but just a list of the parameters and removed unneeded functions. 

I also added a check in `playSound()` within the `countdown.js` to look for whether the `data-play-strong` variable is  a string `false` value.  Otherwise, the library tried to access a URL with: `/false`

<img width="612" alt="Pre-patched countdown JS attempting to play a `/false` URL" src="https://github.com/gadenbuie/countdown/assets/833642/0806f6df-5966-4f74-88ae-792be112f87e">

I'd like to try to get this into the repo this weekend. Given our prior discussions on `prismatic`, what else should we aim for in the PR?

gadenbuie commented on quarto/_extensions/countdown/countdown.lua:
```diff
@@ -0,0 +1,295 @@
+  -- Embed into the document to avoid rendering to disk and, then, embedding a URL.
+  quarto.doc.include_text('in-header', 
+    string.format(
+      "<!-- Countdown Global CSS -->\n<style text='text/css'>:%s {%s}</style>", 
```
Just a small thing that the whole  default selector is `:root`

```suggestion
      "<!-- Countdown Global CSS -->\n<style text='text/css'>%s {%s}</style>", 
```

coatless replied:
Great catch

coatless replied:
FYI this followed the _R_ variant lead:

https://github.com/gadenbuie/countdown/blob/34204099a9b575b72504523f739a51c200bbca5f/r/R/countdown.R#L348

https://github.com/gadenbuie/countdown/blob/34204099a9b575b72504523f739a51c200bbca5f/r/R/countdown.R#L359

So, potentially `.selector = ":root"` and the `:` in the `sprintf()` should be removed.

gadenbuie replied:
🤦‍♂️ oh man, I didn't realize. Thanks for catching that! I'll fix it in a separate PR

gadenbuie replied:
Fixed in R in #47

gadenbuie commented on quarto/_extensions/countdown/countdown.lua:
```diff
@@ -0,0 +1,295 @@
+  end
+
+  -- Determine the selector value
+  local possibleSelector = getOption(options, "selector", "root")
```
```suggestion
  local possibleSelector = getOption(options, "selector", ":root")
```

gadenbuie commented on quarto/_extensions/countdown/countdown.lua:
```diff
@@ -0,0 +1,295 @@
+-- Specify embedded version 
+local countdownEmbeddedVersion = "0.0.1"
+
+-- Only embed resources once if there are multiple timers present 
+local needsToExportDependencies = true
```
I believe Quarto deduplicate dependencies (see first paragraph in https://quarto.org/docs/extensions/lua-api.html#html-dependencies ). So you can declare the same dependency more than once and only one copy will be included. So you could simplify this a little bit and just add the countdown does every time.

coatless replied:
We need the custom catch because we're injecting into the header the global style as part of a text include. To reduce the check from re-happening with the add_html_dependency() and to group the actions, I placed it alongside the header injection.

gadenbuie replied:
Oh I see, that makes sense

gadenbuie reviewed (approved):
This is looking great! Just a few last little things and then feel free to merge when you're ready. Thanks again for your work on this!

gadenbuie commented on quarto/_extensions/countdown/countdown.lua:
```diff
@@ -0,0 +1,295 @@
+    '<span class="countdown-digits seconds">', string.format("%02d", seconds), '</span></code>',
+    '\n </div>'
+  })
+
+  quarto.log.output(rawHtml)
```
We should remove this before merging

```suggestion
```

coatless commented:
@gadenbuie Let's maybe aim to improve the `examples.qmd`. I can port of some examples inside of the `.Rmd`. Would you be up for applying the Garrick Xarginan Styling (TM)?

coatless commented:
I think one other part here is how do we want to handle deployment of the documentation? Do we also want to add a shinylive variant of the R app as well?

gadenbuie commented:
Those are both great ideas! Let's tackle those in follow up PRs -- actually it'd probably be worth creating issues for those, and any other next steps from this PR, so we can talk about them as needed.

I'm very okay with this PR going into main soon; everything works enough for people (or just us) to be able to play with it. It'll be easier to have smaller, targeted PRs from here that fill in some of the gaps.

coatless commented:
@gadenbuie okay; I've split out the issues into tickets for discussion. List:

- #48
- #49
- #50
- #51

Are there any others that should be included?

gadenbuie commented:
Two more follow up tasks:

* #53 
* #54

gadenbuie reviewed (approved):
Thank you again for kicking off this work, I'm excited that countdown will finally be a real, grown-up Quarto extension!

Although we have some more work to do, this is ready for anyone to try out, so I'm going to go ahead and merge.

coatless commented:
@gadenbuie 🎉  👍 

Thanks again for being open for the Quarto & python port 😄
